compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerFilterScope.skiko.kt

```diff
@@ -31,4 +42,31 @@ class PointerFilterScope internal constructor(
     val keyboardModifiers = event.keyboardModifiers
 
     val isMouse: Boolean = event.changes.fastAll { it.type == PointerType.Mouse }
+    val isStylus: Boolean = event.changes.fastAll { it.type == PointerType.Stylus }
+    val isTouch: Boolean = event.changes.fastAll { it.type == PointerType.Touch }
+    val isEraser: Boolean = event.changes.fastAll { it.type == PointerType.Eraser }
+
+    @OptIn(ExperimentalComposeUiApi::class)
+    fun isButtonPressed(button: PointerButton): Boolean {
+        return when (event.type) {
+            PointerEventType.Press,
```

Comment:
A pressed state doesn't depend on the event type, we just can read it from `event.buttons`. Also, better to use `event.buttons.isPressed(button.index)`, and maybe implement `event.buttons.isPressed(button)`, so we can provide here only `buttons`, not a function. Users can call `buttons.isPressed(PointerButton.Primary)` instead of `isButtonPressed(PointerButton.Primary)`

Reply:
The purpose of `isButtonPressed` function is to simplify filtering. (maybe its name is not self-explaining).

Let's consider an example: filter for dragging by Primary button.
It means that drag should be started by Primary button, and Primary button should be pressed until the end.

How can we create such a filter?
```kotlin
{ isMouse && button.isPrimary || !isMouse }
```
This filter is helpful only for press/release events. But during move events, button.isPrimary == false. As you remember we decided to set `PointerEvent.button` value only when that specific button was pressed/released. Therefore users need to somehow configure a filter for `move` events as well. It usually can be done by looking at `PointerEvent.buttons`, but we don't expose it now in `PointerFilterScope` (and I'm not sure if we should). But even if it's exposed, then they would need to apply the same conditional logic `when (event.type) {...}` to determine what should be checked `button` or `buttons`.

Can we rely on `PointerEvent.buttons` for all event types (press, release, move)? 
Yes, at least now. It depends on `PointerInputEventProcessor.process` [implementation](https://github.com/JetBrains/androidx/blob/dcac3bf6cfb514521c69265feee028de62331cd6/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/input/pointer/PointerInputEventProcessor.kt#L63). Currently it ensures that if pointer was pressed within a particular UI subtree, then other parallel branches of UI tree would not see any pointer events even if pointer is moved/pressed on them. 
Maybe we will want to change this logic for desktop, then it will be possible that `PointerEvent.buttons` has `true` values pressed for irrelevant UI subtrees. But it's not the case for now. 


Maybe for now, we can expose `PointerEvent.buttons` in `PointerFilterScope`, but it will bring another question for users: button vs buttons (what to use and when?). We added `button` for clicks filtering, although `buttons` can be used as well (at least with current  PointerInputEventProcessor implementation) Maybe we don't need `button` at all? 


So this comment can be concluded with a question about the following scenario: 
We have a Box (with `pointerInput` modifier) in the center of the screen and nothing else.
1) Move the cursor outside of the Box.
2) Press secondary button.
3) Move the cursor to the center of the Box, while secondary button is pressed.
4) Press primary button.

What's expected result?
Did Box (its `pointerInput{}`) receive Enter event and primary button Press event, or not? 

Currently, it doesn't receive them. 
In my opinion, it should receive them (at least for some scenarios, like moving a file, when hovered directories are highlighted).

Reply:
> Maybe we will want to change this logic for desktop, then it will be possible that PointerEvent.buttons has true values pressed for irrelevant UI subtrees.

> In my opinion, it should receive them

When we press the left button, the component under the cursor captures all the future events, even if we move mouse outside it. I am not sure, if it will feasible and easy to support capturing each button separately (captured component for left button, for right button, etc). Maybe in the future we can support some sort of custom event capturing (i.e. if we press left mouse - we don't capture event focus at all, if we press the right mouse - we capture it).

> Can we rely on PointerEvent.buttons for all event types (press, release, move)?

`PointerButtons` simply represents the current pressed buttons.
If we have 
`PointerButtons(isPrimaryPressed = true, isSecondaryPressed = true)` (no matter if it was press/release/move) then at the moment of the event we have left and right buttons pressed. If we release the second button, then on that release event we have `PointerButtons(isPrimaryPressed = true, isSecondaryPressed = false)`.

> Maybe we will want to change this logic for desktop, then it will be possible that PointerEvent.buttons has true values pressed for irrelevant UI subtrees.

I don't understand why this is related to PointerButtons/PointerButton issue. We don't have true values for irrelevant subtrees because we even don't receive the events there.

> Maybe we don't need button at all?

The alternative to having `button` is to compare the previous buttons with the current ones. 
We probably have to have `PointerEvent.previousButtons` field for that (like `PointerInputChange.previousPress`). And we also don't be able to use the same filter for press/release.

As for checking 'filter' on `move` events, I agree, `button` field won't help us. The name `isButtonPressed` confuse me, because looking at this name I think that it represents the current state of the button, which should be `false` when we release the button (i.e. `isButtonPressed(PointerButton.Primary)` is the same as `buttons.isPrimaryPressed`).

Maybe we don't need to check `filter` on move event? Only on press/release. If we don't miss any press/release, we will start/stop drag the right way:

0. we are checking the drag for the right mouse. filter is `button.isPrimary`.
1. wait for press, which pass the filter
2. accept any move, without checking the filter
3. when we receive the release, which passes the filter, we stop the drag.

Reply:
Done. Current implementation listens to release events and as soon as first release event fits filter conditions, we finish drag.